internal/messagequeue/messagequeue.go

```diff
@@ -489,9 +493,17 @@ func (mq *MessageQueue) pendingWorkCount() int {
 	return mq.bcstWants.pending.Len() + mq.peerWants.pending.Len() + mq.cancels.Len()
 }
 
-func (mq *MessageQueue) extractOutgoingMessage(supportsHave bool) (bsmsg.BitSwapMessage, func()) {
-	// Create a new message
-	msg := bsmsg.New(false)
+func (mq *MessageQueue) extractOutgoingMessage(supportsHave bool) (bsmsg.BitSwapMessage, func([]bsmsg.Entry)) {
+	// For performance reasons we just clear out the fields of the message
+	// instead of creating a new one every time.
+	if mq.msg == nil {
+		// Create a new message
+		mq.msg = bsmsg.New(false)
+	} else {
+		// If there's already a message, reset it
+		mq.msg.Reset(false)
+	}
+	msg := mq.msg
```

Comment:
We could also consider using a `sync.Pool` to share these between message queues and automatically release them when we're done.

Reply:
I think a simpler solution would be to perform the `Reset()` right *after* we've sent the message, so we clear up all the memory for the message, and then we're just keeping around a few pointers per peer.

Reply:
I agree we need to call reset first, but we're still keeping entire _maps_, not just pointers. However, it should be change compared to everything else.